src/ctap/storage.rs

```diff
@@ -187,338 +98,312 @@ impl PersistentStore {
+        // early.
+        let result = iter.map(|(_, credential)| credential).find(|credential| {
+            credential.rp_id == rp_id && credential.credential_id == credential_id
         });
-        if check_cred_protect && user_verification_required {
-            Ok(None)
-        } else {
-            Ok(result)
+        iter_result?;
+        if let Some(cred) = &result {
+            let user_verification_required = cred.cred_protect_policy
+                == Some(CredentialProtectionPolicy::UserVerificationRequired);
+            if check_cred_protect && user_verification_required {
+                return Ok(None);
+            }
         }
+        Ok(result)
     }
 
+    /// Stores or updates a credential.
+    ///
+    /// If a credential with the same RP id and user handle already exists, it is replaced.
     pub fn store_credential(
         &mut self,
-        credential: PublicKeyCredentialSource,
+        new_credential: PublicKeyCredentialSource,
     ) -> Result<(), Ctap2StatusCode> {
-        let key = Key::Credential {
-            rp_id: Some(credential.rp_id.clone()),
-            credential_id: None,
-            user_handle: Some(credential.user_handle.clone()),
-        };
-        let old_entry = self.store.find_one(&key);
-        if old_entry.is_none() && self.count_credentials()? >= MAX_SUPPORTED_RESIDENTIAL_KEYS {
+        // Holds the key of the existing credential if this is an update.
+        let mut old_key = None;
+        // Holds the unordered list of used keys.
+        let mut keys = Vec::new();
+        let mut iter_result = Ok(());
+        let iter = self.iter_credentials(&mut iter_result)?;
+        for (key, credential) in iter {
+            keys.push(key);
+            if credential.rp_id == new_credential.rp_id
+                && credential.user_handle == new_credential.user_handle
+            {
+                if old_key.is_some() {
+                    return Err(Ctap2StatusCode::CTAP2_ERR_VENDOR_INTERNAL_ERROR);
+                }
+                old_key = Some(key);
+            }
+        }
+        iter_result?;
+        if old_key.is_none() && keys.len() >= MAX_SUPPORTED_RESIDENTIAL_KEYS {
             return Err(Ctap2StatusCode::CTAP2_ERR_KEY_STORE_FULL);
         }
-        let credential = serialize_credential(credential)?;
-        let new_entry = StoreEntry {
-            tag: TAG_CREDENTIAL,
-            data: &credential,
-            sensitive: true,
-        };
-        match old_entry {
-            None => self.store.insert(new_entry)?,
-            Some((index, old_entry)) => {
-                debug_assert_eq!(old_entry.tag, TAG_CREDENTIAL);
-                self.store.replace(index, new_entry)?
-            }
+        let key = match old_key {
+            // This is a new credential being added, we need to allocate a free key. We choose the
+            // first available key. This is quadratic in the number of existing keys.
```

Comment:
I was wondering while re-reading: Why does this have to be quadratic? Do you think it's premature optimization to do a custom iteration for linear runtime?

Reply:
It was initially to avoid premature optimization, but actually if we only look at worst time and memory complexity, there's indeed no reason to not do linear. Done.